# Coursera_Capstone_Project<br>
The final part of week 3 on which i working with clusters of Toronto


In [29]:
import requests # library to handle requests
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner
import random # library for random number generation


!pip install geopy
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 
    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize


! pip install folium==0.5.0
import folium # plotting library

print('Folium installed')
print('Libraries imported.')

     |████████████████████████████████| 79 kB 9.4 MB/s  eta 0:00:01
  Created wheel for folium: filename=folium-0.5.0-py3-none-any.whl size=76240 sha256=0bfc59869903e7b7e3ee99705c9acc7aa0adee0282fd822f0615e6481463deba
  Stored in directory: /tmp/wsuser/.cache/pip/wheels/b2/2f/2c/109e446b990d663ea5ce9b078b5e7c1a9c45cca91f377080f8
Successfully built folium
  Attempting uninstall: folium
    Found existing installation: folium 0.12.1
    Uninstalling folium-0.12.1:
      Successfully uninstalled folium-0.12.1
Folium installed
Libraries imported.


<b>Importing the last data set with all the features from the previous steps

In [30]:
def __iter__(self): return 0

# @hidden_cell
# The following code accesses a file in your IBM Cloud Object Storage. It includes your credentials.
# You might want to remove those credentials before you share the notebook.
client_dc634226746e43d0bacc8ef81627305b = ibm_boto3.client(service_name='s3',
    ibm_api_key_id='fIrt3A0GR7UBCSqU5DPpbiIHX7jiT9y_t7gH9T1BHtwz',
    ibm_auth_endpoint="https://iam.cloud.ibm.com/oidc/token",
    config=Config(signature_version='oauth'),
    endpoint_url='https://s3.eu-geo.objectstorage.service.networklayer.com')

body = client_dc634226746e43d0bacc8ef81627305b.get_object(Bucket='courseracapstoneproject-donotdelete-pr-zspibbsqs5lgj8',Key='toronto_Coordinates.csv')['Body']
# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

df = pd.read_csv(body)
df.head()


Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

<b> Choosing to work only with Boroughs containing the word Toronto.

In [31]:
df_tor=df[df['Borough'].str.contains("Toronto")].reset_index(drop=True)
df_tor.head()

Postal Code           Borough                                Neighbourhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031

<b>Creating a map which includes marks for every Brrough location containing the word Toronto.<br>
    Each marker has a popup text which contains all the Neighbourhoods and the Borough at the end.

In [32]:
# for the city Toronto, latitude and longtitude are manually extracted via google search
toronto_latitude = 43.6532; toronto_longitude = -79.3832
map_toronto = folium.Map(location = [toronto_latitude, toronto_longitude], zoom_start = 10.7)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_tor['Latitude'], df_tor['Longitude'], df_tor['Borough'], df_tor['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7).add_to(map_toronto)  
    

map_toronto

<b> Storing the Forsquare Credentials 

In [33]:
CLIENT_ID = 'ET0WGFUJUCKRKKNBG4AUPDUMV24FNN3A53UNITL3YPHNV4ZQ' # your Foursquare ID
CLIENT_SECRET = 'GRLHUEH2OG2T5I5MNPYFL0B5RP4ZD3I3T3BFTBDGLOA35KVC' # your Foursquare Secret
ACCESS_TOKEN = 'XBLN52XN45AX2BMUPBWRNN0DTHJCHIJVXCDQDDKEVXMKQVGM' # your FourSquare Access Token
VERSION = '20180604'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: ET0WGFUJUCKRKKNBG4AUPDUMV24FNN3A53UNITL3YPHNV4ZQ
CLIENT_SECRET:GRLHUEH2OG2T5I5MNPYFL0B5RP4ZD3I3T3BFTBDGLOA35KVC


#### Let's again assume that you are staying at the Fairmont Royal York Hotel. So let's start by converting the Fairmont Royal York Hotel's address to its latitude and longitude coordinates.


In [48]:
address = '100 Front St W, Toronto, ON M5J 1E3, Canada'

geolocator = Nominatim(user_agent="foursquare_agent")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print(latitude, longitude)

43.64590915 -79.3813637083646


## 1. Search for a specific venue category <br>
#### Now, let's assume that it is lunch time, and you are craving Italian food. So, let's define a query to search for Italian food that is within 500 metres from the Conrad Hotel.


In [50]:
search_query = 'Italian'
radius = 500
print(search_query + ' .... OK!')
url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&oauth_token={}&v={}&query={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude,ACCESS_TOKEN, VERSION, search_query, radius, LIMIT)
url

Italian .... OK!


'https://api.foursquare.com/v2/venues/search?client_id=ET0WGFUJUCKRKKNBG4AUPDUMV24FNN3A53UNITL3YPHNV4ZQ&client_secret=GRLHUEH2OG2T5I5MNPYFL0B5RP4ZD3I3T3BFTBDGLOA35KVC&ll=43.64590915,-79.3813637083646&oauth_token=XBLN52XN45AX2BMUPBWRNN0DTHJCHIJVXCDQDDKEVXMKQVGM&v=20180604&query=Italian&radius=500&limit=30'

In [51]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '600877613858e2550c42e4d0'},
 'notifications': [{'type': 'notificationTray', 'item': {'unreadCount': 0}}],
 'response': {'venues': [{'id': '5b897e92db1d81002c91df8c',
    'name': 'Fabbrica Rustic Italian',
    'location': {'address': '66 Wellington St W',
     'lat': 43.647161,
     'lng': -79.381691,
     'labeledLatLngs': [{'label': 'display',
       'lat': 43.647161,
       'lng': -79.381691}],
     'distance': 141,
     'postalCode': 'M5K 1E7',
     'cc': 'CA',
     'city': 'Toronto',
     'state': 'ON',
     'country': 'Canada',
     'formattedAddress': ['66 Wellington St W',
      'Toronto ON M5K 1E7',
      'Canada']},
    'categories': [{'id': '4bf58dd8d48988d110941735',
      'name': 'Italian Restaurant',
      'pluralName': 'Italian Restaurants',
      'shortName': 'Italian',
      'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/italian_',
       'suffix': '.png'},
      'primary': True}],
    'referralId': 'v-1611167585',
 

#### Get relevant part of JSON and transform it into a _pandas_ dataframe


In [55]:
# assign relevant part of JSON to venues
venues = results['response']['venues']

# tranform venues into a dataframe
dataframe = pd.json_normalize(venues)
dataframe.head()

id                              name  \
0  5b897e92db1d81002c91df8c           Fabbrica Rustic Italian   
1  581cad6a7c74e15859a6f890  Scaddabush Italian Kitchen & Bar   
2  4b7d777df964a52043c02fe3                        Sauté Rosé   
3  4b846dd4f964a520dc3431e3                  Little Anthony's   

                                          categories    referralId  hasPerk  \
0  [{'id': '4bf58dd8d48988d110941735', 'name': 'I...  v-1611167585    False   
1  [{'id': '4bf58dd8d48988d110941735', 'name': 'I...  v-1611167585    False   
2  [{'id': '4bf58dd8d48988d110941735', 'name': 'I...  v-1611167585    False   
3  [{'id': '4bf58dd8d48988d110941735', 'name': 'I...  v-1611167585    False   

                  location.address  location.lat  location.lng  \
0               66 Wellington St W     43.647161    -79.381691   
1  200 Front Street W., Unit #G001     43.644737    -79.385355   
2               55 University Ave.     43.646547    -79.384029   
3               121 Richmond St. W     43.650296    -79.384513   

                             location.labeledLatLngs  location.distance  \
0  [{'label': 'display', 'lat': 43.647161, 'lng':...                141   
1  [{'label': 'display', 'lat': 43.6447367776608,...                346   
2  [{'label': 'display', 'lat': 43.64654731418263...                226   
3  [{'label': 'display', 'lat': 43.65029624519052...                550   

  location.postalCode location.cc location.city location.state  \
0             M5K 1E7          CA       Toronto             ON   
1             M5V 3J1          CA       Toronto             ON   
2             M5J 2H7          CA       Toronto             ON   
3             M5H 2K1          CA       Toronto             ON   

  location.country                          location.formattedAddress  \
0           Canada   [66 Wellington St W, Toronto ON M5K 1E7, Canada]   
1           Canada  [200 Front Street W., Unit #G001 (at Simcoe St...   
2           Canada  [55 University Ave. (in Markel Building), Toro...   
3           Canada  [121 Richmond St. W (at York St.), Toronto ON ...   

  location.crossStreet venuePage.id  
0                  NaN          NaN  
1         at Simcoe St          NaN  
2   in Markel Building          NaN  
3          at York St.     75877692

In [56]:
# keep only columns that include venue name, and anything that is associated with location
filtered_columns = ['name', 'categories'] + [col for col in dataframe.columns if col.startswith('location.')] + ['id']
dataframe_filtered = dataframe.loc[:, filtered_columns]

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
dataframe_filtered['categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# clean column names by keeping only last term
dataframe_filtered.columns = [column.split('.')[-1] for column in dataframe_filtered.columns]

dataframe_filtered

name          categories  \
0           Fabbrica Rustic Italian  Italian Restaurant   
1  Scaddabush Italian Kitchen & Bar  Italian Restaurant   
2                        Sauté Rosé  Italian Restaurant   
3                  Little Anthony's  Italian Restaurant   

                           address        lat        lng  \
0               66 Wellington St W  43.647161 -79.381691   
1  200 Front Street W., Unit #G001  43.644737 -79.385355   
2               55 University Ave.  43.646547 -79.384029   
3               121 Richmond St. W  43.650296 -79.384513   

                                      labeledLatLngs  distance postalCode  cc  \
0  [{'label': 'display', 'lat': 43.647161, 'lng':...       141    M5K 1E7  CA   
1  [{'label': 'display', 'lat': 43.6447367776608,...       346    M5V 3J1  CA   
2  [{'label': 'display', 'lat': 43.64654731418263...       226    M5J 2H7  CA   
3  [{'label': 'display', 'lat': 43.65029624519052...       550    M5H 2K1  CA   

      city state country                                   formattedAddress  \
0  Toronto    ON  Canada   [66 Wellington St W, Toronto ON M5K 1E7, Canada]   
1  Toronto    ON  Canada  [200 Front Street W., Unit #G001 (at Simcoe St...   
2  Toronto    ON  Canada  [55 University Ave. (in Markel Building), Toro...   
3  Toronto    ON  Canada  [121 Richmond St. W (at York St.), Toronto ON ...   

          crossStreet                        id  
0                 NaN  5b897e92db1d81002c91df8c  
1        at Simcoe St  581cad6a7c74e15859a6f890  
2  in Markel Building  4b7d777df964a52043c02fe3  
3         at York St.  4b846dd4f964a520dc3431e3

#### Let's visualize the Italian restaurants that are nearby


In [58]:
venues_map = folium.Map(location=[latitude, longitude], zoom_start=13) # generate map centred around the Conrad Hotel

# add a red circle marker to represent the Conrad Hotel
folium.CircleMarker(
    [latitude, longitude],
    radius=10,
    color='red',
    popup='Fairmont Royal York Hotel',
    fill = True,
    fill_color = 'red',
    fill_opacity = 0.6
).add_to(venues_map)

# add the Italian restaurants as blue circle markers
for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered.lng, dataframe_filtered.categories):
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        color='blue',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(venues_map)

# display map
venues_map

## 2. Explore a Given Venue
### A. Let's explore the closests restaurants --Fabbrica Rustic Italian, Scaddabush Italian Kitchen & Bar, Sauté Rosé, Little Anthony's
####  We will get the ratings of the closest Italian restaurants so we can choose the best one

In [84]:
venue_id = ['5b897e92db1d81002c91df8c','581cad6a7c74e15859a6f890','4b7d777df964a52043c02fe3','4b846dd4f964a520dc3431e3'] # ID of Harry's Italian Pizza Bar

url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&oauth_token={}&v={}'.format(venue_id[0], CLIENT_ID, CLIENT_SECRET,ACCESS_TOKEN, VERSION)
result = requests.get(url).json()
    
print(result['response']['venue'].keys())
result['response']['venue']

dict_keys(['id', 'name', 'contact', 'location', 'canonicalUrl', 'categories', 'verified', 'stats', 'price', 'likes', 'like', 'dislike', 'ok', 'allowMenuUrlEdit', 'beenHere', 'specials', 'photos', 'reasons', 'hereNow', 'createdAt', 'tips', 'shortUrl', 'timeZone', 'listed', 'seasonalHours', 'pageUpdates', 'inbox', 'parent', 'hierarchy', 'attributes'])


{'id': '5b897e92db1d81002c91df8c',
 'name': 'Fabbrica Rustic Italian',
 'contact': {},
 'location': {'address': '66 Wellington St W',
  'lat': 43.647161,
  'lng': -79.381691,
  'labeledLatLngs': [{'label': 'display',
    'lat': 43.647161,
    'lng': -79.381691}],
  'postalCode': 'M5K 1E7',
  'cc': 'CA',
  'city': 'Toronto',
  'state': 'ON',
  'country': 'Canada',
  'formattedAddress': ['66 Wellington St W', 'Toronto ON M5K 1E7', 'Canada']},
 'canonicalUrl': 'https://foursquare.com/v/fabbrica-rustic-italian/5b897e92db1d81002c91df8c',
 'categories': [{'id': '4bf58dd8d48988d110941735',
   'name': 'Italian Restaurant',
   'pluralName': 'Italian Restaurants',
   'shortName': 'Italian',
   'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/italian_',
    'suffix': '.png'},
   'primary': True},
  {'id': '4bf58dd8d48988d1ca941735',
   'name': 'Pizza Place',
   'pluralName': 'Pizza Places',
   'shortName': 'Pizza',
   'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/

### B. Get the venue's overall rating


In [85]:
try:
    print(result['response']['venue']['rating'])
except:
    print('This venue has not been rated yet.')

This venue has not been rated yet.


In [87]:
url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&oauth_token={}&v={}'.format(venue_id[1], CLIENT_ID, CLIENT_SECRET,ACCESS_TOKEN, VERSION)
result = requests.get(url).json()
    
try:
    print(result['response']['venue']['rating'])
except:
    print('This venue has not been rated yet.')

7.2


In [88]:
url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&oauth_token={}&v={}'.format(venue_id[2], CLIENT_ID, CLIENT_SECRET,ACCESS_TOKEN, VERSION)
result = requests.get(url).json()
    
try:
    print(result['response']['venue']['rating'])
except:
    print('This venue has not been rated yet.')

This venue has not been rated yet.


In [89]:
url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&oauth_token={}&v={}'.format(venue_id[3], CLIENT_ID, CLIENT_SECRET,ACCESS_TOKEN, VERSION)
result = requests.get(url).json()
try:
    print(result['response']['venue']['rating'])
except:
    print('This venue has not been rated yet.')

6.6


 So we can conclude that the highest rated reataurant is <b> Scaddabush Italian Kitchen & Bar with rating 7.2